crates/walrus-sui/src/client/retry_client.rs

```diff
@@ -1554,42 +1617,93 @@ impl RetriableRpcClient {
                 .await
-            })
+                .map_err(RetriableClientError::from)
+        }
+
+        let request = |client: Arc<FallibleRpcClient>| {
+            let inner_request = retry_rpc_errors(
+                self.get_extended_strategy(),
+                move || make_request(client.clone(), self.request_timeout),
+                self.metrics.clone(),
+                "get_latest_checkpoint_summary",
+            );
+            Box::pin(inner_request)
         };
 
         self.client.with_failover(request).await
     }
 
     /// Gets the object with the given ID.
     pub async fn get_object(&self, id: ObjectID) -> Result<Object, RetriableClientError> {
-        let request = |client: Arc<RpcClient>| {
-            Box::pin(async move {
-                retry_rpc_errors(
-                    self.get_extended_strategy(),
-                    || Self::handle_rpc_timeout(self.request_timeout, client.get_object(id)),
-                    self.metrics.clone(),
-                    "get_object",
+        async fn make_request(
+            client: Arc<FallibleRpcClient>,
+            id: ObjectID,
+            request_timeout: Duration,
+        ) -> Result<Object, RetriableClientError> {
+            client
+                .call(
+                    |rpc_client| async move { rpc_client.get_object(id).await },
+                    request_timeout,
                 )
                 .await
-            })
+                .map_err(RetriableClientError::from)
+        }
+
+        let request = |client: Arc<FallibleRpcClient>| {
+            let inner_request = retry_rpc_errors(
+                self.get_extended_strategy(),
+                move || make_request(client.clone(), id, self.request_timeout),
+                self.metrics.clone(),
+                "get_object",
+            );
+            Box::pin(inner_request)
         };
 
         self.client.with_failover(request).await
     }
+
+    /// Updates the last known height update time
+    fn update_last_height(&self) {
+        self.last_height_update
+            .store(Instant::now(), Ordering::Relaxed);
+    }
+
+    /// Gets the exponential window size based on time since last height update
+    fn get_window_size(&self) -> u64 {
+        let time_since_last_height = self.last_height_update.load(Ordering::Relaxed).elapsed();
+        let seconds_since_last_height = time_since_last_height.as_secs();
+
+        // For the first minute, disable fallback
+        if seconds_since_last_height < 60 {
+            return 0;
+        }
+
+        // After first minute, start exponential growth
+        // Subtract 60 to start counting from when exponential growth begins
+        let seconds_since_growth_start = seconds_since_last_height - 60;
+
+        // Base window size of 50 checkpoints (10 seconds worth at ~5 checkpoints/second)
+        let base_window = 50;
+
+        // Exponential growth after first minute (assuming 5 checkpoints/second), this gives us:
```

Comment:
Note: in general, we get about 4.2 checkpoints per seconds.

Reply:
not needed anymore